File: leetcode/12_integer-to-roman/lc12.cpp
class Solution {
public:
    string intToRoman(int num) {
        vector<int> value{1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1};
        vector<string> token{"M", "CM", "D", "CD", "C", "XC", "L", "XL", "X", "IX", "V", "IV", "I"};
        string roman;
        for (int i = 0; i < value.size(); ++i) {
            int v = value[i];
            while (num >= v) {
                roman += token[i];
                num -= v;
            }
        }
        return roman;
    }
};

File: leetcode/12_integer-to-roman/lc12.py
class Solution:
    def intToRoman(self, num: int) -> str:
        token = [['CM', 'XC', 'IX'],
                 ['D', 'L', 'V'],
                 ['C', 'X', 'I'],
                 ['CD', 'XL', 'IV']]
        res = ''
        cur1, s1, cur2, s2, s3 = 1000, 900, 500, 400, 100
        res += 'M' * (num // cur1)
        for i in range(3):
            if num % cur1 >= s1:
                res += token[0][i]
            elif num % cur1 >= cur2:
                res += token[1][i] + token[2][i] * (num % cur2 // s3)
            elif num % cur1 >= s2:
                res += token[3][i]
            else:
                res += token[2][i] * (num % cur2 // s3)
            cur1, s1, cur2, s2, s3 = cur1 // 10, s1 // 10, cur2 // 10, s2 // 10, s3 // 10
        return res
